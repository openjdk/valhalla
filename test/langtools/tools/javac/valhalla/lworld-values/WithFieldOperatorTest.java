/*
 * @test /nodynamiccopyright/
 * @summary Verify that various errors related to __WithField operator are caught.
 * @compile/fail/ref=WithFieldOperatorTest.out -XDallowWithFieldOperator -XDrawDiagnostics -XDdev WithFieldOperatorTest.java
 */

public class WithFieldOperatorTest {
    static int xs;
    int ifld;
    class Y {}
    public final primitive class V { int x = 10; }

    public final primitive class X {

        final int x;
        final V v;

        X() {
            x = 10;
            v = V.default;
        }
        
        X getX(int xVal, WithFieldOperatorTest wfot) {
            X x = X.default;
            x = __WithField(new Y(), null);  // not a variable at all.
            x = __WithField(wfot.xs, 10); // not an instance field.
            x = __WithField(wfot.ifld, 10); // not a field of value type
            x = __WithField(xVal, xVal); // not a field
            x = __WithField(this, this); // not a field
            x = __WithField(X.this, this); // not a field
            x = __WithField(x.x, 12.0); // float cannot be assigned to int
            x = __WithField(x.v, null); // null cannot be assigned to value 
            return x;
        }
    }
}

class WithFieldOperatorTest_aux {
    void foo(WithFieldOperatorTest.X x) {
        x = __WithField(x.x, 10); // outside the nest
    }
}
