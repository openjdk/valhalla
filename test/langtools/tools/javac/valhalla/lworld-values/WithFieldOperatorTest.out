WithFieldOperatorTest.java:25:29: compiler.err.unexpected.type: kindname.variable, kindname.value
WithFieldOperatorTest.java:26:33: compiler.err.primitive.class.instance.field.expected.here
WithFieldOperatorTest.java:27:33: compiler.err.primitive.class.instance.field.expected.here
WithFieldOperatorTest.java:28:29: compiler.err.primitive.class.instance.field.expected.here
WithFieldOperatorTest.java:29:29: compiler.err.cant.assign.val.to.this
WithFieldOperatorTest.java:30:30: compiler.err.cant.assign.val.to.this
WithFieldOperatorTest.java:31:34: compiler.err.prob.found.req: (compiler.misc.possible.loss.of.precision: double, int)
WithFieldOperatorTest.java:32:34: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: compiler.misc.type.null, WithFieldOperatorTest.V)
WithFieldOperatorTest.java:40:26: compiler.err.cant.assign.val.to.final.var: x
9 errors
